### PR TITLE
Better Wandb logging

### PR DESCRIPTION
**Pull request by faresobeid**
Removed logging of total tokens, and duplicate logging in both orchestrator and trainer
Added logging of logprob error (recompute vs inference logprobs), solve_none, solve_all, ...

### CODE REVIEW

File: src/prime_rl/trainer/train.py [hunk 1]
```diff
@@ -189,8 +189,18 @@ def train(config: TrainerConfig):
...
 
-                    logprobs = compute_logprobs(logprob_model, input_ids, position_ids, temperature)
-                    micro_batch["logprobs"] = logprobs.to("cpu")
+                    recomputed_logprobs = compute_logprobs(logprob_model, input_ids, position_ids, temperature)
+
+                    original_logprobs = micro_batch["logprobs"].to(recomputed_logprobs.device)
+
+                    diff_mask = micro_batch["loss_mask"][:, 1:].to(recomputed_logprobs.device)
+
+                    abs_diff = (recomputed_logprobs - original_logprobs).abs()
```
Comment (samsja) [inline src/prime_rl/trainer/train.py]:
```suggestion
                    abs_diff_recomputed_logprobs = (recomputed_logprobs - original_logprobs).abs()
```

File: src/prime_rl/trainer/train.py [hunk 2]
```diff
@@ -189,8 +189,18 @@ def train(config: TrainerConfig):
...
+                    abs_diff_sum = (torch.exp(abs_diff) * diff_mask).sum()
```
Comment (samsja) [inline src/prime_rl/trainer/train.py]:
```suggestion
                    abs_diff = 
                    abs_diff_recomputed_logprobs_sum = (torch.exp(recomputed_logprobs - original_logprobs).abs()) * diff_mask).sum()
```

File: src/prime_rl/trainer/train.py [hunk 3]
```diff
@@ -248,6 +258,10 @@ def train(config: TrainerConfig):
             loss_metrics["loss/loss"] += loss.detach().clone().float()
             loss_metrics["loss/entropy"] += entropy.detach().clone().float()
             loss_metrics["loss/importance_ratio"] += importance_ratio.detach().clone().float()
+            loss_metrics["loss/clipped_ratio"] += clip_token_count.detach().clone().float()
+
+            abs_diff_sum = micro_batch.get("abs_logprob_diff_sum", torch.tensor(0.0))
```
Comment (samsja) [inline src/prime_rl/trainer/train.py]:
```suggestion
            abs_diff_recomputed_logprobs_sum = micro_batch.get("abs_logprob_diff_sum", torch.tensor(0.0))
```

**Review by samsja — COMMENTED** [review]
looks good, I added couple of comments, generaly its better to have longer variable name that are more descritive

File: src/prime_rl/orchestrator/orchestrator.py [hunk 4]
```diff
@@ -215,7 +215,7 @@ async def orchestrate(config: OrchestratorConfig):
         rewards = outputs["reward"]  # TODO: Align naming between prime-rl <> verifiers
 
         # Compute advantages
-        advantages = compute_advantages(rewards, config.rollouts_per_prompt)
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/orchestrator.py:218]: im not a huge fan of returning logging only metrics as a tuple from a function (we have the same for the loss funcs which i think we should also change), this seems like it can quickly become convoluted, but it's more code style so i can change later

File: src/prime_rl/orchestrator/orchestrator.py [hunk 5]
```diff
@@ -285,17 +287,20 @@ async def orchestrate(config: OrchestratorConfig):
...
         # Log rewards metrics to monitor
         reward_metrics = {
             "reward/reward": np.mean(rewards),
-            "reward/reward_std": np.std(rewards),
-            "reward/advantage": np.mean(advantages),
-            "reward/advantage_std": np.std(advantages),
-            "reward/advantage_zero_ratio": np.mean(np.array(advantages) == 0.0),
+            "reward/solve_none_ratio": solve_none_ratio,
+            "reward/solve_all_ratio": solve_all_ratio,
+            "reward/effective_batch_size_ratio": effective_batch_size_ratio,
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/orchestrator.py]: i think we can drop the ratio suffix here. by default, we always compute ratios over the batch. otherwise, it would also be the `reward_ratio` etc.

File: src/prime_rl/orchestrator/orchestrator.py [hunk 6]
```diff
@@ -230,6 +230,9 @@ async def orchestrate(config: OrchestratorConfig):
         throughput = num_tokens / (generate_completions_time)
         avg_seq_length = num_tokens / config.batch_size
 
+        seq_lengths = np.array([len(p) + len(c) for p, c in zip(prompt_tokens, completion_tokens)])
+        problem_avg_seqlens = seq_lengths.reshape(-1, config.rollouts_per_prompt).mean(axis=-1)
+
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/orchestrator.py]: i like it. it's more general but then let's remove the `avg_seq_lengh` computation from before and compute it from the seq_lengths array as `avg_seq_length = seq_lengths.mean()`

File: src/prime_rl/orchestrator/orchestrator.py [hunk 7]
```diff
@@ -273,7 +276,6 @@ async def orchestrate(config: OrchestratorConfig):
 
         # Log progress metrics to monitor
         progress_metrics = {
-            "progress/orchestrator/total_tokens": progress.total_tokens,
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/orchestrator.py:276]: oh okay so the point of basically all progress metrics is that we might want to use either of these as our "x-axis". the default is obviously the step, but it might be valuable to also view progress as a function of samples or tokens to determine sample/ compute efficiency of a run, so i think we should keep this in

File: src/prime_rl/trainer/train.py [hunk 8]
```diff
@@ -248,6 +248,8 @@ def train(config: TrainerConfig):
             loss_metrics["loss/loss"] += loss.detach().clone().float()
             loss_metrics["loss/entropy"] += entropy.detach().clone().float()
             loss_metrics["loss/importance_ratio"] += importance_ratio.detach().clone().float()
+            loss_metrics["loss/clipped_ratio"] += clip_token_count.detach().clone().float()
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/train.py]: same here, i would remove the ratio. i think for importance ratio it makes sense to keep tho bc its actually the logprob ratio right?

File: src/prime_rl/trainer/loss.py [hunk 9]
```diff
@@ -91,14 +91,14 @@ def grpo_loss_clip(
...
+    clip_token_count = _masked_sum(is_clipped, loss_mask)
 
     if highest_entropy_percentage < 1.0:
         loss_mask = highest_entropy_mask(logits, loss_mask, highest_entropy_percentage)
 
     loss = _masked_sum(per_token_loss, loss_mask)
-
-    return loss, clip_ratio_tensor
+    ratio = _masked_sum(coef_2, loss_mask)
+    return loss, ratio, clip_token_count
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py]: is the ratio returned here the same as for ratio loss?

File: src/prime_rl/trainer/train.py [hunk 10]
```diff
@@ -319,15 +321,6 @@ def train(config: TrainerConfig):
...
         logger.success(step_message)
 
-        # Log progress metrics
-        progress_metrics = {
-            "progress/train/total_tokens": progress.total_tokens,
-            "progress/train/total_samples": progress.total_samples,
-            "progress/train/step": progress.step,  # Shared W&B axis
-            "step": progress.step,
-        }
-        monitor.log(progress_metrics)
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/train.py:329]: okay, so after rethinking, i think this is actually fine to do. my concern before was that the orchestrator and trainer both have their own step definition 0,1,2,... which "point to" different model checkpoints. at async level 2 the orchestrator's steps 0,1,2 are from model checkpoint 0 (base model) and the values are getting logged like this. but the orchestrator also logs the the ckpt step which is correctly at 0,0,0 for the first 2 steps. if not both the trainer and the orchestrator log this key we cannot use it as a shared axis i believe. i think we can switch the axis just for the orchestrator plots tho and then all is aligned so should be good

File: src/prime_rl/orchestrator/orchestrator.py [hunk 11]
```diff
@@ -273,7 +276,6 @@ async def orchestrate(config: OrchestratorConfig):
...
             "progress/orchestrator/total_samples": progress.total_samples,
             "progress/orchestrator/epoch": progress.epoch,
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/orchestrator.py]: lets delete the orchestrator from these then tho
```suggestion
            "progress/total_samples": progress.total_samples,
            "progress/epoch": progress.epoch,
```

**Review by mikasenghaas — APPROVED** [review]

File: src/prime_rl/trainer/train.py (hunk 8 above)
Comment (mikasenghaas) [reply src/prime_rl/trainer/train.py]: *the ratio suffix

File: src/prime_rl/trainer/loss.py (hunk 9 above)
Comment (faresobeid) [reply src/prime_rl/trainer/loss.py]: wdym?

File: src/prime_rl/environments/registry.py [hunk 12]
```diff
@@ -39,6 +39,37 @@ def correct_answer_reward_func(completion, answer, **kwargs) -> float:
...
+
+    rubric = vf.Rubric(
+        funcs=[
+            correct_answer_reward_func,
+        ],
+        weights=[1.0],
+    )
+
+    vf_env = vf.SingleTurnEnv(dataset=train_dataset, rubric=rubric)
+    return vf_env
```
Comment (mikasenghaas) [inline src/prime_rl/environments/registry.py]: this is a separate pr?

File: src/prime_rl/trainer/loss.py (hunk 9 above)
Comment (mikasenghaas) [reply src/prime_rl/trainer/loss.py]: both the clipped and ratio loss return a ratio, is it equivalent in meaning?

File: src/prime_rl/orchestrator/orchestrator.py [hunk 13]
```diff
@@ -230,6 +230,9 @@ async def orchestrate(config: OrchestratorConfig):
         throughput = num_tokens / (generate_completions_time)
         avg_seq_length = num_tokens / config.batch_size
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/orchestrator.py]: delete this?

File: src/prime_rl/trainer/loss.py (hunk 9 above)
Comment (samsja) [reply src/prime_rl/trainer/loss.py]: Yes it is

File: src/prime_rl/environments/registry.py (hunk 12 above)
Comment (faresobeid) [reply src/prime_rl/environments/registry.py]: ya mistake, fixed it

**Review by samsja — APPROVED** [review]
Lfgtm 🔥
